token/core/common/interop/pledge/validator_pledge.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/identity"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/interop/pledge"
+	"github.com/pkg/errors"
+)
+
+func IssuePledgeValidate[P driver.PublicParameters, T driver.Output, TA driver.TransferAction, IA driver.IssueAction, DS driver.Deserializer](ctx *common.Context[P, T, TA, IA, DS]) error {
+	for k := range ctx.IssueAction.GetMetadata() {
+		ctx.CountMetadataKey(k)
+	}
+	return nil
+}
+
+func TransferPledgeValidate[P driver.PublicParameters, T driver.Output, TA driver.TransferAction, IA driver.IssueAction, DS driver.Deserializer](ctx *common.Context[P, T, TA, IA, DS]) error {
+	for _, in := range ctx.InputTokens {
+		id, err := identity.UnmarshalTypedIdentity(in.GetOwner())
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal owner of input token")
+		}
+		if id.Type == pledge.ScriptType {
+			if len(ctx.InputTokens) != 1 || len(ctx.TransferAction.GetOutputs()) != 1 {
+				return errors.Errorf("invalid transfer action: a pledge script only transfers the ownership of a token")
+			}
+			out := ctx.TransferAction.GetOutputs()[0]
+			sender, err := identity.UnmarshalTypedIdentity(ctx.InputTokens[0].GetOwner())
+			if err != nil {
+				return err
+			}
+			script := &pledge.Script{}
+			err = json.Unmarshal(sender.Identity, script)
+			if err != nil {
+				return err
+			}
+			if time.Now().Before(script.Deadline) {
+				return errors.New("cannot reclaim pledge yet: wait for timeout to elapse.")
+			}
+
+			key, err := constructMetadataKey(ctx.TransferAction)
+			if err != nil {
+				return errors.Wrap(err, "failed constructing metadata key")
+			}
+
+			var metadataKey string
+			if out.IsRedeem() {
+				metadataKey = pledge.RedeemPledgeKey + key
+			} else {
+				metadataKey = pledge.MetadataReclaimKey + key
+				if !script.Sender.Equal(out.GetOwner()) {
+					return errors.New("recipient of token does not correspond to sender of reclaim request")
+				}
+			}
+
+			v, ok := ctx.TransferAction.GetMetadata()[metadataKey]
+			if !ok {
+				return errors.Errorf("metadata key not found [%s]", metadataKey)
+			}
+			if v == nil {
+				return errors.Errorf("empty metadata for key [%s]", metadataKey)
+			}
+			ctx.CountMetadataKey(metadataKey)
+		}
+	}
+
+	for _, out := range ctx.TransferAction.GetOutputs() {
+		if out.IsRedeem() {
+			continue
+		}
+		owner, err := identity.UnmarshalTypedIdentity(out.GetOwner())
+		if err != nil {
+			return err
+		}
+		if owner.Type == pledge.ScriptType {
+			script := &pledge.Script{}
+			err = json.Unmarshal(owner.Identity, script)
+			if err != nil {
+				return err
+			}
+			if script.Deadline.Before(time.Now()) {
+				return errors.Errorf("pledge script is invalid: expiration date has already passed")
+			}
+			v, ok := ctx.TransferAction.GetMetadata()[pledge.MetadataKey+script.ID]
+			if !ok {
+				return errors.Errorf("empty metadata for pledge script with identifier %s", script.ID)
+			}
```

Comment:
What information does this metadata convey?

Reply:
I see that this is used in a `IDExists` in the scanner to find if there is a token pledged with a given `ID`. That method though it does not seem to be used.

Reply:
Going through the rest of the code, this seems to enforce to enforce the uniqueness of `PledgeID`.